File: personalizatio-sdk/src/main/java/com/personalizatio/stories/views/StoryDialog.java
package com.personalizatio.stories.views;

import android.app.Dialog;
import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.Rect;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager2.widget.ViewPager2;

import com.personalizatio.R;
import com.personalizatio.stories.models.Story;

import java.util.HashMap;
import java.util.List;

final public class StoryDialog extends Dialog implements PullDismissLayout.Listener {

	private final List<Story> stories;
	private final HashMap<Integer, StoryView> storyViews = new HashMap<>();
	private final ViewPagerAdapter adapter;
	private ViewPager2 mViewPager;

	private final Runnable completeListener;
	private final int startPosition;
	private final StoriesView storiesView;

	public interface OnProgressState {
		void onState(boolean running);
	}

	public StoryDialog(StoriesView stories_view, List<Story> stories, int start_position, Runnable completeListener) {
		super(stories_view.getContext(), android.R.style.Theme_Translucent_NoTitleBar);
		requestWindowFeature(Window.FEATURE_NO_TITLE);
		setContentView(R.layout.dialog_stories);
		Window window = getWindow();
		WindowManager.LayoutParams wlp = window.getAttributes();

		wlp.gravity = Gravity.CENTER;
		wlp.flags &= ~WindowManager.LayoutParams.FLAG_BLUR_BEHIND;
		window.setAttributes(wlp);
		window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

		this.storiesView = stories_view;
		this.stories = stories;
		this.startPosition = start_position;
		adapter = new ViewPagerAdapter(running -> mViewPager.setUserInputEnabled(running));

		this.completeListener = completeListener;

		setOnCancelListener(dialog -> {
			for(int i = 0; i < stories.size(); i++) {
				StoryView holder = getHolder(i);
				if( holder != null ) {
					holder.release();
				}
			}
		});

		setupViews();
	}

	private void setupViews() {
		((PullDismissLayout) findViewById(R.id.pull_dismiss_layout)).setListener(this);
		ImageButton closeImageButton = findViewById(R.id.imageButton);
		closeImageButton.setOnClickListener(v -> {
			onDismissed();
		});
		closeImageButton.setColorFilter(Color.parseColor(storiesView.getSettings().close_color));
		mViewPager = findViewById(R.id.view_pager);
		mViewPager.setClipToPadding(false);
		mViewPager.setClipChildren(false);
		mViewPager.setOffscreenPageLimit(1);

		//Превью слайдов только в горизонтальном экране
		if( getContext().getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE ) {
			DisplayMetrics dm = getContext().getResources().getDisplayMetrics();
			float nextItemVisiblePx = (dm.widthPixels - dm.heightPixels * 9f / 16) / 2f;
			float currentItemHorizontalMarginPx = nextItemVisiblePx + getContext().getResources().getDimension(R.dimen.viewpager_current_item_horizontal_margin);
			float pageTranslationX = nextItemVisiblePx + currentItemHorizontalMarginPx;
			mViewPager.setPageTransformer((page, position) -> {
				page.setTranslationX(-pageTranslationX * position);
				page.setScaleY(1 - (0.25f * Math.abs(position)));
				page.setScaleX(1 - (0.25f * Math.abs(position)));
				page.setAlpha(1 - (0.7f * Math.abs(position)));
			});
			mViewPager.addItemDecoration(new RecyclerView.ItemDecoration() {
				@Override
				public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
					super.getItemOffsets(outRect, view, parent, state);
					outRect.right = (int) currentItemHorizontalMarginPx;
					outRect.left = (int) currentItemHorizontalMarginPx;
				}
			});
		}

		mViewPager.setAdapter(adapter);
		//Хак, чтобы не срабатывал onPageSelected при открытии первой кампании
		mViewPager.setCurrentItem(startPosition == 0 ? stories.size() : 0, false);
		//Устанавливаем позицию
		mViewPager.setCurrentItem(startPosition, false);
		mViewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {

			@Override
			public void onPageSelected(int position) {
				for(int i = 0; i < stories.size(); i++ ) {
					StoryView holder = getHolder(i);
					if( holder != null ) {
						if( i != position ) {
							holder.stopStories();
						}
					}
				}
				StoryView holder = getHolder(position);
				if( holder != null ) {
					holder.startStories();
				}
			}

			@Override
			public void onPageScrollStateChanged(int state) {
				if( state == ViewPager2.SCROLL_STATE_IDLE ) {
					onReleased();
				}
			}
		});
	}

	private StoryView getCurrentHolder() {
		return getHolder(mViewPager.getCurrentItem());
	}

	private StoryView getHolder(int position) {
		return storyViews.get(position);
	}

	public void onDetachedFromWindow() {
		//При закрытии диалогового окна, возвращаем все метки в исходное
		for( Story story : stories ) {
			for (var i = 0; i < story.getSlidesCount(); i++){
				story.getSlide(i).setPrepared(false);
			}
		}
	}

	@Override
	public void onDismissed() {
		cancel();
	}

	@Override
	public void onReleased() {
		StoryView holder = getCurrentHolder();
		if( holder != null ) {
			holder.resume();
		}
	}

	@Override
	public boolean onShouldInterceptTouchEvent() {
		return false;
	}

	static class PagerHolder extends RecyclerView.ViewHolder {
		public PagerHolder(@NonNull StoryView itemView) {
			super(itemView);
		}
	}

	class ViewPagerAdapter extends RecyclerView.Adapter<PagerHolder> {

		private final OnProgressState state_listener;

		public ViewPagerAdapter(OnProgressState state_listener) {
			this.state_listener = state_listener;
		}

		@NonNull
		@Override
		public PagerHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			return new PagerHolder(new StoryView(storiesView, state_listener));
		}

		@Override
		public void onViewAttachedToWindow(@NonNull PagerHolder holder) {
			if( holder.getLayoutPosition() == mViewPager.getCurrentItem() ) {
				super.onViewAttachedToWindow(holder);
				StoryView storyView = (StoryView) holder.itemView;
				storyView.startStories();
			}
		}

		@Override
		public void onBindViewHolder(@NonNull PagerHolder holder, int position) {
			StoryView view = (StoryView) holder.itemView;
			storyViews.put(position, view);
			view.setStory(stories.get(position), () -> {
				if( position >= stories.size() || position + 1 >= stories.size() ) {
					cancel();
				} else {
					mViewPager.setCurrentItem(position + 1);
				}
				completeListener.run();
			}, () -> {
				if( position == 0 ) {
					cancel();
				} else {
					mViewPager.setCurrentItem(position - 1);
				}
			});
		}

		@Override
		public int getItemCount() {
			return stories.size();
		}
	}
}

File: personalization-sdk/src/main/kotlin/com/personalization/stories/views/StoryDialog.kt
package com.personalization.stories.views

import android.app.Dialog
import android.content.res.Configuration
import android.graphics.Color
import android.graphics.Rect
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.ImageButton
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.personalization.R
import com.personalization.stories.StoryState
import com.personalization.stories.models.Story
import kotlin.math.abs

class StoryDialog(
    private val storiesView: StoriesView,
    private val stories: List<Story>,
    private val startPosition: Int,
    private val completeShowStory: () -> Unit,
    private val cancelShowStory: () -> Unit
) : Dialog(
    storiesView.context,
    android.R.style.Theme_Translucent_NoTitleBar
), PullDismissLayout.Listener {

    private val storyViews = HashMap<Int, StoryView>()
    private val adapter: ViewPagerAdapter
    private var mViewPager: ViewPager2

    fun interface OnStoryStateListener {
        fun onStoryStateChanged(storyState: StoryState)
    }

    init {
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.dialog_stories)
        val window = window
        val wlp = window!!.attributes

        wlp.gravity = Gravity.CENTER
        wlp.flags = wlp.flags and WindowManager.LayoutParams.FLAG_BLUR_BEHIND.inv()
        window.attributes = wlp
        window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)

        mViewPager = findViewById(R.id.view_pager)

        adapter = ViewPagerAdapter { storyState ->
            when (storyState) {
                StoryState.RUNNING -> mViewPager.isUserInputEnabled = true
                StoryState.PAUSE -> mViewPager.isUserInputEnabled = false
                StoryState.CLOSE -> cancel()
            }
        }

        setOnCancelListener {
            for (i in stories.indices) {
                getHolder(i)?.release()
            }

            cancelShowStory()
        }

        setupViews()
    }

    private fun setupViews() {
        (findViewById<View>(R.id.pull_dismiss_layout) as PullDismissLayout).listener = this

        val closeImageButton = findViewById<ImageButton>(R.id.imageButton)
        closeImageButton.setOnClickListener { onDismissed() }
        closeImageButton.setColorFilter(Color.parseColor(storiesView.settings.close_color))

        mViewPager.clipToPadding = false
        mViewPager.clipChildren = false
        mViewPager.offscreenPageLimit = 1

        if (context.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            val dm = context.resources.displayMetrics
            val nextItemVisiblePx = (dm.widthPixels - dm.heightPixels * 9f / 16) / 2f
            val currentItemHorizontalMarginPx =
                nextItemVisiblePx + context.resources.getDimension(R.dimen.viewpager_current_item_horizontal_margin)
            val pageTranslationX = nextItemVisiblePx + currentItemHorizontalMarginPx
            mViewPager.setPageTransformer { page: View, position: Float ->
                val absPosition = abs(position.toDouble())
                val verticalScale = (1 - (0.25f * absPosition)).toFloat()
                page.translationX = -pageTranslationX * position
                page.scaleY = verticalScale
                page.scaleX = verticalScale
                page.alpha = (1 - (0.7f * absPosition)).toFloat()
            }
            mViewPager.addItemDecoration(object : ItemDecoration() {
                override fun getItemOffsets(
                    outRect: Rect,
                    view: View,
                    parent: RecyclerView,
                    state: RecyclerView.State
                ) {
                    super.getItemOffsets(outRect, view, parent, state)
                    val margin = currentItemHorizontalMarginPx.toInt()
                    outRect.right = margin
                    outRect.left = margin
                }
            })
        }

        mViewPager.adapter = adapter

        //Hack to prevent onPageSelected from triggering when opening the first campaign
        mViewPager.setCurrentItem(if (startPosition == 0) stories.size else 0, false)

        mViewPager.setCurrentItem(startPosition, false)
        mViewPager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                for (i in stories.indices) {
                    if (i != position) {
                        getHolder(i)?.stopStories()
                    }
                }
                getHolder(position)?.startStories()
            }

            override fun onPageScrollStateChanged(state: Int) {
                if (state == ViewPager2.SCROLL_STATE_IDLE) {
                    onReleased()
                }
            }
        })
    }

    private val currentHolder: StoryView?
        get() = getHolder(mViewPager.currentItem)

    private fun getHolder(position: Int): StoryView? {
        return storyViews[position]
    }

    override fun onDetachedFromWindow() {
        for (story in stories) {
            for (i in 0 until story.slidesCount) {
                story.getSlide(i).isPrepared = false
            }
        }
    }

    override fun onDismissed() {
        cancel()
    }

    override fun onReleased() {
        currentHolder?.resume()
    }

    override fun onShouldInterceptTouchEvent(): Boolean {
        return false
    }

    internal class PagerHolder(itemView: StoryView) : RecyclerView.ViewHolder(itemView)

    internal inner class ViewPagerAdapter(
        private val storyStateListener: OnStoryStateListener
    ) :
        RecyclerView.Adapter<PagerHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PagerHolder {
            return PagerHolder(StoryView(storiesView, storyStateListener))
        }

        override fun onViewAttachedToWindow(holder: PagerHolder) {
            if (holder.layoutPosition == mViewPager.currentItem) {
                super.onViewAttachedToWindow(holder)
                val storyView = holder.itemView as StoryView
                storyView.startStories()
            }
        }

        override fun onBindViewHolder(holder: PagerHolder, position: Int) {
            val view = holder.itemView as StoryView
            storyViews[position] = view
            view.setStory(stories[position], {
                if (position >= stories.size || position + 1 >= stories.size) {
                    cancel()
                } else {
                    mViewPager.currentItem = position + 1
                }

                completeShowStory()
            }, {
                if (position == 0) {
                    cancel()
                } else {
                    mViewPager.currentItem = position - 1
                }
            })
        }

        override fun getItemCount(): Int {
            return stories.size
        }
    }
}
